zebra-state/src/service/finalized_state/disk_db.rs

```diff
@@ -514,6 +495,62 @@ impl DiskDb {
             .map(|(key, value)| (K::from_bytes(key), V::from_bytes(value)))
     }
 
+    /// Returns the RocksDB ReadOptions with a lower and upper bound for a range.
+    fn zs_iter_opts<R>(range: &R) -> ReadOptions
+    where
+        R: RangeBounds<Vec<u8>>,
+    {
+        let mut opts = ReadOptions::default();
+        let (lower_bound, upper_bound) = Self::zs_iter_bounds(range);
+
+        if let Some(bound) = lower_bound {
+            opts.set_iterate_lower_bound(bound);
+        };
+
+        if let Some(bound) = upper_bound {
+            opts.set_iterate_upper_bound(bound);
+        };
+
+        opts
+    }
+
+    /// Returns a lower and upper iterate bounds for a range.
+    ///
+    /// Note: Since upper iterate bounds are always exclusive in RocksDB, this method
+    ///       will increment the upper bound by 1 if the end bound of the provided range
+    ///       is inclusive, or will return an upper bound of `None` if the end bound of a
+    ///       provided range is inclusive and already the max key for that column family.
+    fn zs_iter_bounds<R>(range: &R) -> (Option<Vec<u8>>, Option<Vec<u8>>)
+    where
+        R: RangeBounds<Vec<u8>>,
+    {
+        use std::ops::Bound::*;
+
+        let lower_bound = match range.start_bound() {
+            Included(bound) | Excluded(bound) => Some(bound.clone()),
+            Unbounded => None,
+        };
+
+        let upper_bound = match range.end_bound().cloned() {
+            Included(mut bound) => {
+                // Skip adding an upper bound if every byte is u8::MAX, or
+                // increment the last byte in the upper bound that is less than u8::MAX,
+                // and clear any bytes after it to increment the big-endian number this
+                // Vec represents to RocksDB.
+                let is_max_key = bound.iter_mut().rev().all(|v| {
+                    *v = v.wrapping_add(1);
+                    v == &0
+                });
+
+                (!is_max_key).then_some(bound)
```

Comment:
What happens if the key is variable-length?

Reply:
It would sometimes (<1%) fail to put an upper bound on iterators, which is still constrained by `take_while` but might unnecessarily read-ahead.

Its not an issue for the sprout/history tree change where the key size changes across db format versions, but if key structs had variable length serialized representations, like if heights 0-255 were just 1 byte, this would need something like a method in `IntoDisk` for checking the max serialization size of a struct.

I think we should add support for variable sized keys here when it's needed.

Reply:
Zebra's data access patterns aren't random, so I'm not sure if that `<1%` probability will hold under deliberate attack, or with certain key formats. Here are two examples:

#### Attackers using maximum keys

In an attack or overload scenario, if the keys depend on mempool or block data, then many of them could be at the maximum value.

But this is unlikely to happen with the current UTXO key format, because the height comes first, and it is always less than the maximum. So if we decide to accept this risk, we should document it. (But see my alternative solution below.)

#### Zebra using maximum keys

In the sprout and history tree changes, the new key is `()`, which has an empty array as its serialization. This is the maximum value for an empty array. 

An earlier PR fixed this issue by using `zs_get()` to access keys, rather than an iterator.

But without that fix, every iterator that used `()` would read some tombstones, because the byte array increment code always returns `None` for `[0; u8]`. This would temporarily reduce performance after the format change, until those tombstones were compacted. (Sometimes this doesn't happen for a long time.)

#### Proposed Solution

If the key is the maximum for its current length, add an extra `0x01` byte to the front of the array. This will give you the  minimal bound.

Reply:
> If the key is the maximum for its current length, add an extra 0x01 byte to the front of the array. This will give you the minimal bound.

That's a good idea.

Applied in https://github.com/ZcashFoundation/zebra/pull/7731/commits/fcf171f10be906f0c77041401ad1b59d5adadbaf.